### PR TITLE
[ADD] account: suspense account management

### PR DESCRIPTION
**Pull request by william-andre**
Task 1930403

A suspense account is an account in the general ledger that temporarily stores any transactions for which there is uncertainty about the account in which they should be recorded. Once the accounting staff investigates and clarifies the purpose of this type of transaction, it shifts the transaction out of the suspense account and into the correct account(s). An entry into a suspense account may be a debit or a credit.

The Chosen Odoo approach (macro) for this case is
1) from bank statement -> reconciliation widget -> reconciliation model
2) create move that have a dedicated checkbox
3) be able to reopen the bank statement line "like if there is no linked account move" and finally reconcile the transaction with the good invoice/bill (we can reprocess account move "to check")
4) when we reprocess, Instead of creating an account move : Check if there is an "to check" existing linked account move and remove all the account move lines then replace  them by the good ones --> we want to keep the sequence

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/account/models/account_invoice.py [hunk 1]
```diff
@@ -354,6 +377,7 @@ def _compute_payments(self):
         related='partner_id.commercial_partner_id', store=True, readonly=True,
         help="The commercial entity that will be used on Journal Entries for this invoice")
 
+    match_to_check = fields.Boolean(compute='_get_match_to_check', groups='account.group_account_invoice')
```
Comment (qdp-odoo) [inline addons/account/models/account_invoice.py]: rename field into 'edition_mode_available'  (+change function name accordingly)

File: addons/account/models/account_invoice.py [hunk 2]
```diff
@@ -122,6 +122,29 @@ def _compute_residual(self):
...
+            ('full_reconcile_id', '=', False),
+        ]
+        if self.type == 'out_invoice':
+            domain.append(('balance', '=', -self.residual))
+        elif self.type == 'in_invoice':
+            domain.append(('balance', '=', self.residual))
+        elif self.type == 'out_refund':
+            domain.append(('balance', '=', self.residual))
+        elif self.type == 'in_refund':
+            domain.append(('balance', '=', -self.residual))
```
[the remark this answers is not in the document]
Comment (qdp-odoo) [reply addons/account/models/account_invoice.py]: as in action_reconcile_to_check() you're doing `mapped('statement_line_id').ids`, I think you need to this to the domain as well, otherwise it just gonna be weird as you may end up seeing the message that won't be opening any statement line in the reconciliation widget

File: addons/account/models/account_journal_dashboard.py [hunk 3]
```diff
@@ -308,6 +310,22 @@ def action_open_reconcile(self):
...
+        action_context.update({'mode': 'to_check'})
+        action_context.update({'statement_line_ids': ids})
+        return {
+            'type': 'ir.actions.client',
+            'tag': 'bank_statement_reconciliation_view',
+            'context': action_context,
+        }
+    
+    def to_check_ids(self):
+        statement_line_ids = self.env['account.move'].search([('journal_id', 'in', self.ids), ('to_check', '=', True)]).mapped('line_ids.statement_line_id')
```
Comment (qdp-odoo) [inline addons/account/models/account_journal_dashboard.py]: domain here and on the invoice aren't the same but definitively should (full_reconcile_id clause is missing)

File: addons/account/models/account_reconcile_model.py [hunk 4]
```diff
@@ -625,6 +626,23 @@ def _apply_rules(self, st_lines, excluded_ids=None, partner_map=None):
...
+                if self._context.get('mode') == 'to_check':
+                    query = '''
+                        SELECT aml.id FROM (
+                            SELECT aml.id, aml.full_reconcile_id
+                            FROM account_move_line aml
+                            JOIN account_move am ON aml.move_id = am.id
+                            WHERE am.to_check = True
+                            AND aml.full_reconcile_id > 0
+                            AND aml.id IN %s) as t
+                        JOIN account_move_line as aml ON aml.full_reconcile_id = t.full_reconcile_id AND aml.id != t.id;
```
Comment (qdp-odoo) [inline addons/account/models/account_reconcile_model.py]: these conditions on full_reconcile_is are dubious... I don't get why it is not simply `line.journal_entry_ids.filtered(lambda l: l.account_id.id not in accounts_set_on_journal)`

File: addons/account/models/reconciliation_widget.py [hunk 5]
```diff
@@ -39,10 +40,13 @@ def process_bank_statement_line(self, st_line_ids, data):
...
                 st_line.write({'partner_id': datum['partner_id']})
 
-            st_line.with_context(ctx).process_reconciliation(
+            account_move = st_line.with_context(ctx).process_reconciliation(
                 datum.get('counterpart_aml_dicts', []),
                 payment_aml_rec,
                 datum.get('new_aml_dicts', []))
+            
+            if datum.get('to_check') is not None:
+                account_move.write({'to_check': datum['to_check']})
```
Comment (qdp-odoo) [inline addons/account/models/reconciliation_widget.py]: avoid that write() by creating the account.move directly with the good value

File: addons/account/models/reconciliation_widget.py [hunk 6]
```diff
@@ -153,26 +161,46 @@ def get_bank_statement_line_data(self, st_line_ids, excluded_ids=None):
         return results
 
     @api.model
-    def get_bank_statement_data(self, bank_statement_ids):
+    def get_bank_statement_data(self, bank_statement_ids, search_str=False):
```
Comment (qdp-odoo) [inline addons/account/models/reconciliation_widget.py]: what is this change? everything done in this function doesn't seem related to the task... can you explain?

File: addons/account/models/reconciliation_widget.py [hunk 7]
```diff
@@ -417,14 +445,22 @@ def _domain_move_lines(self, search_str):
...
             :param partner_id:
             :param excluded_ids:
             :param search_str:
         """
+        to_check = self._context.get('mode') == 'to_check'
+        
+        if to_check:
+            to_check_excluded = self.env['account.move.line'].search([
+                ('move_id.to_check', '=', True),
+                ('full_reconcile_id', '=', False),
```
Comment (qdp-odoo) [inline addons/account/models/reconciliation_widget.py]: it's the third time you need to search account.move.line that are 'toc_check', this lightens a signal that it needs a function returning the domain to use for the search. Doesn't it?

File: addons/account/models/reconciliation_widget.py [hunk 8]
```diff
@@ -529,8 +574,8 @@ def _prepare_move_lines(self, move_lines, target_currency=False, target_date=Fal
 
             debit = line.debit
             credit = line.credit
-            amount = line.amount_residual
-            amount_currency = line.amount_residual_currency
+            amount = line.balance if to_check else line.amount_residual
+            amount_currency = line.amount_currency if to_check else line.amount_residual_currency
```
Comment (qdp-odoo) [inline addons/account/models/reconciliation_widget.py]: o_O

not sure at all, but this needs testing for confirmation

File: addons/account/models/account_invoice.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (william-andre) [reply addons/account/models/account_invoice.py]: @avw asked to remove the constraint and changed the specs considering what we should expect as an expense account

File: addons/account/models/reconciliation_widget.py (hunk 8 above)
Comment (william-andre) [reply addons/account/models/reconciliation_widget.py]: Maybe this isn't how I should do it, but amount_residual is 0 because the line is already reconciled. I will call button_cancel_reconciliation right after, but before it is done, I need to know the balance of the line

**Comment by qdp-odoo** [comment]
@william-andre please add tests

File: addons/account/models/reconciliation_widget.py (hunk 6 above)
Comment (qdp-odoo) [reply addons/account/models/reconciliation_widget.py]:
```python
def get_bank_statement_data(self, bank_statement_line_ids, search_str=False):
    bank_statements = self.env['account.bank.statement.line'].browse(bank_statement_line_ids).mapped('statement_id')
```
that would be much cleaner...

File: addons/account/views/account_invoice_view.xml [hunk 9]
```diff
@@ -424,7 +428,11 @@
                 <div groups="account.group_account_invoice" class="alert alert-info" role="alert" style="margin-bottom:0px;" attrs="{'invisible': [('has_outstanding','=',False)]}">
                     You have <bold><a class="alert-link" href="#outstanding" role="button">outstanding payments</a></bold> for this customer. You can allocate them to mark this invoice as paid.
                 </div>
+                <div groups="account.group_account_invoice" class="alert alert-info" role="alert" style="margin-bottom:0px;" attrs="{'invisible': [('edition_mode_available','=',False)]}">
+                    You have suspense account moves that match this invoice. <bold><button class="alert-link" type="object" name="action_reconcile_to_check" role="button" string="Check them" style="padding: 0 0 1px 0;"/></bold> to mark this invoice as paid.
```
Comment (qdp-odoo) [inline addons/account/views/account_invoice_view.xml]: @csnauwaert can you check that button? The style is weird and I've tried to change with `<a action="action_reconcile_to_check">Check them</a>`, the display is good then but it's not clickable anymore... -_-

**Comment by qdp-odoo** [comment]
@csnauwaert can you just check the js part, on Monday please?

File: addons/account/views/account_invoice_view.xml (hunk 9 above)
Comment (csnauwaert) [reply addons/account/views/account_invoice_view.xml]: @qdp-odoo did you try adding a type= object to the a tag? <a type="object" action="action_reconcile_to_check">Check them</a>
It should works with the type and the action.

Comment (william-andre) [reply addons/account/views/account_invoice_view.xml]: @csnauwaert It doesn't work

File: addons/account/models/account_move.py [hunk 10]
```diff
@@ -317,8 +318,14 @@ def action_post(self):
...
     def button_cancel(self):
+        AccountMoveLine = self.env['account.move.line']
+        excluded_move_ids = []
+
+        if self._context.get('edition_mode'):
+            excluded_move_ids = AccountMoveLine.search(AccountMoveLine._get_domain_for_edition_mode()).mapped('move_id').ids
+
         for move in self:
-            if not move.journal_id.update_posted:
+            if not move.journal_id.update_posted and move.id not in excluded_move_ids:
```
Comment (antonylesuisse) [inline addons/account/models/account_move.py]: Check the condition here move.to_check == True, then the lines to avoid loading the ids.

Comment (qdp-odoo) [reply addons/account/models/account_move.py]: I changed to 
```diff
        if self._context.get('edition_mode'):
-            excluded_move_ids = AccountMoveLine.search(AccountMoveLine._get_domain_for_edition_mode()).mapped('move_id').ids
+            excluded_move_ids = AccountMoveLine.search(AccountMoveLine._get_domain_for_edition_mode() + [('id', 'in', self.ids)]).mapped('move_id').ids
        for move in self:
            if not move.journal_id.update_posted and move.id not in excluded_move_ids:
                ...
```

File: addons/account/static/tests/reconciliation_tests.js [hunk 11]
```diff
@@ -1857,5 +1858,65 @@ QUnit.module('account', {
...
+            var def = $.Deferred();
+            def.abort = function () {};
+            event.data.callback(def);
+        });
+
+        widget.$("button.o_reconcile.btn.btn-primary:first").click();
+        assert.ok(true, "No error in reconciliation");
+
+        clientAction.destroy();
+    });
```
Comment (qdp-odoo) [inline addons/account/static/tests/reconciliation_tests.js]: @csnauwaert can you check this test please?

File: addons/account/static/src/js/reconciliation/reconciliation_model.js [hunk 12]
```diff
@@ -1027,6 +1054,13 @@ var StatementModel = BasicModel.extend({
                     return data.write_off ? self.quickCreateProposition(line.handle, data.model_id) : true;
                 })
             );
+            // If still no partner set, take the one from context, if it exists
```
Comment (csnauwaert) [inline addons/account/static/src/js/reconciliation/reconciliation_model.js]: This seems wrong..
You probably wanted to put that part after the last 'then' 2 lines above.

```
.then(function(){
                    return data.write_off ? self.quickCreateProposition(line.handle, data.model_id) : true;
                })
.then(function(){
       if (!line.st_line.partner_id && self.context.partner_id && self.context.partner_name) {
                return self.changePartner(line.handle, {
                    'id': self.context.partner_id,
                    'display_name': self.context.partner_name,
                }, true);
            }
           return true;
})
```

File: addons/account/static/src/js/reconciliation/reconciliation_action.js [hunk 13]
```diff
@@ -32,6 +32,9 @@ var StatementAction = AbstractAction.extend({
         load_more: '_onLoadMore',
         reload: 'reload',
     },
+    events: {
+        'change .reconciliation_search_input': '_onSearch',
```
Comment (csnauwaert) [inline addons/account/static/src/js/reconciliation/reconciliation_action.js]: Is it possible that this event just reload the page without changing anything?

File: addons/account/static/tests/reconciliation_tests.js (hunk 11 above)
Comment (csnauwaert) [reply addons/account/static/tests/reconciliation_tests.js]: seems correct to me

File: addons/account/static/src/js/reconciliation/reconciliation_model.js (hunk 12 above)
Comment (william-andre) [reply addons/account/static/src/js/reconciliation/reconciliation_model.js]: It seems to be doing the job just fine though, what would be the problem with this?

**Comment by qdp-odoo** [comment]
everything looks ok to me, can you make a single commit, rebase with master and check why the runbot is red (maybe not related to your stuff)

thanks

**Comment by qdp-odoo** [comment]
@robodoo r+
